### PR TITLE
Add json output for crc version

### PR DESCRIPTION
**Pull request by guillaumerose**
Continuation of #1227

### CODE REVIEW

**Comment by anjannath** [comment]
Commit message in https://github.com/code-ready/crc/pull/1429/commits/6ae8ca217c8e4f786be03f880cc9f25b0a5137c8 has typos

**Comment by guillaumerose** [comment]
thanks fixed!

File: cmd/crc/cmd/version.go [hunk 1]
```diff
@@ -11,6 +12,7 @@ import (
 )
 
 func init() {
+	versionCmd.Flags().StringVarP(&outputFormat, "output", "o", "", "Output format. One of: json")
```
[the remark this answers is not in the document]
Comment (anjannath) [reply cmd/crc/cmd/version.go]: Aha, It was there in the previous PR for status command, https://github.com/code-ready/crc/pull/1402#discussion_r459288573

i did check other tools like `odo` and `oc`, and those also don't mention the default output format.

[the remark this answers is not in the document]
Comment (cfergeau) [reply cmd/crc/cmd/version.go]: hmm yeah it's a bit weird in `oc`, the default is apparently `''`, and one can use `-o ''`. There is no explicit `-o xxx` which would correspond to the default in at least `oc version`

File: cmd/crc/cmd/version.go [hunk 2]
```diff
@@ -19,18 +21,23 @@ var versionCmd = &cobra.Command{
...
-		if err := runPrintVersion(os.Stdout); err != nil {
+		if err := runPrintVersion(os.Stdout, defaultVersion(), outputFormat); err != nil {
 			exit.WithMessage(1, err.Error())
 		}
 	},
 }
 
-func runPrintVersion(writer io.Writer) error {
-	return printVersion(writer, defaultVersion())
+func runPrintVersion(writer io.Writer, version *version, outputFormat string) error {
```
[the remark this answers is not in the document]
Comment (guillaumerose) [reply cmd/crc/cmd/version.go:30]: done :+1:

File: cmd/crc/cmd/version.go [hunk 3]
```diff
@@ -19,18 +21,23 @@ var versionCmd = &cobra.Command{
...
+		encoder := json.NewEncoder(writer)
+		encoder.SetIndent("", "  ")
+		return encoder.Encode(version)
+	}
+	return printVersion(writer, version)
 }
 
-func printVersion(writer io.Writer, v *version) error {
-	for _, line := range v.lines() {
+func printVersion(writer io.Writer, version *version) error {
```
[the remark this answers is not in the document]
Comment (guillaumerose) [reply cmd/crc/cmd/version.go]: done :+1:

File: cmd/crc/cmd/version.go [hunk 4]
```diff
@@ -19,18 +21,23 @@ var versionCmd = &cobra.Command{
...
 
-func runPrintVersion(writer io.Writer) error {
-	return printVersion(writer, defaultVersion())
+func runPrintVersion(writer io.Writer, version *version, outputFormat string) error {
+	if outputFormat == jsonFormat {
+		encoder := json.NewEncoder(writer)
+		encoder.SetIndent("", "  ")
+		return encoder.Encode(version)
+	}
+	return printVersion(writer, version)
```
[the remark this answers is not in the document]
Comment (guillaumerose) [reply cmd/crc/cmd/version.go]: done :+1:

File: cmd/crc/cmd/json.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+func render(obj prettyPrintable, writer io.Writer, outputFormat string) error {
+	switch outputFormat {
+	case jsonFormat:
+		encoder := json.NewEncoder(writer)
+		encoder.SetIndent("", "  ")
+		return encoder.Encode(obj)
+	default:
+		return obj.prettyPrintTo(writer)
```
Comment (cfergeau) [inline cmd/crc/cmd/json.go:31]: or
```
"":
    return obj.prettyPrintTo(writer)
default:
    return fmt.Errorf("Invalid format: %s", outputFormat")
```
?

File: cmd/crc/cmd/status.go [hunk 6]
```diff
@@ -68,38 +61,31 @@ func runStatus(writer io.Writer, client client, cacheDir, outputFormat string) e
...
 		CacheDir:         cacheDir,
-	}
-
-	if outputFormat == jsonFormat {
-		encoder := json.NewEncoder(writer)
-		encoder.SetIndent("", "  ")
-		return encoder.Encode(status)
-	}
-	return printStatus(writer, status)
+	}, writer, outputFormat)
```
Comment (cfergeau) [inline cmd/crc/cmd/status.go:72]: Fwiw, I find
```
status := Status { ...}
return render(&status, writer, outputFormat)
```
easier to read

Comment (guillaumerose) [reply cmd/crc/cmd/status.go]: done !

File: cmd/crc/cmd/json.go (hunk 5 above)
Comment (guillaumerose) [reply cmd/crc/cmd/json.go:32]: done :)

**Pull request merged**